### PR TITLE
Generic protein purification

### PR DESCRIPTION
**Pull request by eskila**
See https://docs.google.com/document/d/1vhHu6qoDQmV3SWbxzxxoH41j31-1q8JASVP8tmtPUnQ/edit?usp=sharing

### CODE REVIEW

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py [hunk 1]
```diff
@@ -0,0 +1,639 @@
...
+    total_vol_after_bead_adddn = round(total_lysis_vol + bead_volume, 1)
+    if use_NaCl:
+        total_vol_after_bead_adddn = round(total_vol_after_bead_adddn +
+                                           nacl_first_step_vol, 1)
+
+    n_columns = math.ceil(n_samples/8)
+
+    def str_to_bool(str):
+        return True if str.lower().strip() == "true" else False
+
```
Comment (chazmanc) [inline protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:129]: let's chat about this

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py [hunk 2]
```diff
@@ -0,0 +1,639 @@
...
+    do_SDS_step = str_to_bool(do_SDS_step)
+    do_DNAse_step = str_to_bool(do_DNAse_step)
+    use_NaCl = str_to_bool(use_NaCl)
+
+    # Error checking
+
+    # If DNAse I is being added from a tube we need to make sure that
+    # a) we have a single pipette loaded and b) that dnaseI_vol > pipette min.
+    # volume
+    if (("single" not in small_pipette_lname and
```
Comment (chazmanc) [inline protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:141]: 👍

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py [hunk 3]
```diff
@@ -0,0 +1,639 @@
...
+
+    Note: if you are loading a thermocycler, you do not need to specify
+    a slot number - thermocyclers will always occupy slots 7, 8, 10, and 11.
+
+    For all other modules, you can load them on slots 1, 3, 4, 6, 7, 9, 10.
+
+    '''
+
+    # import pdb; pdb.set_trace()
+    # load labware
```
Comment (chazmanc) [inline protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:163]: please remove dead code like this throughout

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py [hunk 4]
```diff
@@ -0,0 +1,639 @@
...
+
+        Other parameters: See class
+        opentrons.protocol_api.contexts.InstrumentContext
+        """
+        try:
+            pip.transfer(volume, source, dest, kwargs)
+        except protocol_api.labware.OutOfTipsError:
+            ctx.pause("\n\nReplace empty tip racks for {}\n".format(pip.name))
+            pip.reset_tipracks()
+            pip.transfer(volume, source, dest, kwargs)
```
Comment (chazmanc) [inline protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:397]: you have a `pick_up` function that essentially does the same thing...

**Review by chazmanc — COMMENTED** [review]
python looks good to me - just add the readme and fields.json

File: protocols/generic_protein_purification/README.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
+# Generic protein purification protocol
+
+### Author
+[Opentrons](https://opentrons.com/)
+
+## Categories
+* Broader Category
+	* Subcategory (can be the name of a kit when applicable)
```
Comment (chazmanc) [inline protocols/generic_protein_purification/README.md:8]: update categories

File: protocols/generic_protein_purification/README.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+* `Number of times to mix the bead solution before transferring beads to the samples` Optional number of times to use the large pipette to mix the bead solution before transferring the bead solution to the samples
+* `SDS buffer volume to elute proteins with (optional: this only matters if you decide to elute with SDS buffer)` In the final step you can choose whether to elute with elution buffer or SDS buffer if you want to run a gel on the samples, this parameter specifies the volume in microliter of SDS buffer.
+---
+
+### Modules
+* [Temperature Module (GEN2)](https://shop.opentrons.com/collections/hardware-modules/products/tempdeck)
+* [Magnetic Module (GEN2)](https://shop.opentrons.com/collections/hardware-modules/products/magdeck)
+
+### Labware
+* [Labware Library](https://labware.opentrons.com/)
```
Comment (chazmanc) [inline protocols/generic_protein_purification/README.md:57]: give some examples of labware used

File: protocols/generic_protein_purification/README.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+### Pipettes
+* [Opentrons Pipettes](https://opentrons.com/pipettes/)
+
+### Reagents
+* [MagneHis™ Protein Purification System](https://www.promega.com/products/protein-purification/protein-purification-kits/magnehis-protein-purification-system/?catNum=V8500#protocols)
```
Comment (chazmanc) [inline protocols/generic_protein_purification/README.md:63]: Protein Purification Kit (for example, MagneHis...)

File: protocols/generic_protein_purification/fields.json [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+    ]
+  },
+  {
+    "type": "dropDown",
+    "label": "Reagent reservoir",
+    "name": "reservoir_lname",
+    "options": [
+      {"label": "NEST 12 Well Reservoir 15 mL", "value": "nest_12_reservoir_15ml" },
+      {"label": "USA Scientific 12 Well Reservoir 22 mL", "value": "usascientific_12_reservoir_22ml" },
+      {"label": "Custom", "value": "custom" }
```
Comment (chazmanc) [inline protocols/generic_protein_purification/fields.json:79]: remove `custom` as an option for now

File: protocols/generic_protein_purification/fields.json [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+    ]
+  },
+  {
+    "type": "dropDown",
+    "label": "Destination plate",
+    "name": "destination_plate_lname",
+    "options": [
+      {"label": "", "value": "" },
+      {"label": "", "value": "" },
```
Comment (chazmanc) [inline protocols/generic_protein_purification/fields.json:88]: fill in and remove `custom`

File: protocols/generic_protein_purification/fields.json [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+  },
+  {
+    "type": "dropDown",
+    "label": "Tube rack (optional)",
+    "name": "tube_rack_lname",
+    "options": [
+      { "label": "", "value": "" },
+      { "label": "", "value": "" }
+    ]
+  },
```
Comment (chazmanc) [inline protocols/generic_protein_purification/fields.json:123]: delete

File: protocols/generic_protein_purification/fields.json [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+    "default": 5
+  },
+  {
+    "type": "float",
+    "label": "SDS buffer volume to elute proteins with (optional: this only matters if you decide to elute with SDS buffer)",
+    "name": "sds_buffer_vol",
+    "default": 30
+  }
+]
+* `Magnet engagment time for bead binding`
```
Comment (chazmanc) [inline protocols/generic_protein_purification/fields.json:246]: delete

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+                        "reservoir_lname": "nest_12_reservoir_15ml",
+                        "destination_plate_lname":"biorad_96_wellplate_200ul_pcr",
+                        "sample_plate_lname":"usascientific_96_wellplate_2.4ml_deep",
+                        "mag_mod_lname":"magnetic module gen2",
+                        "mag_engage_time":5,
+                        "tube_rack_lname":"opentrons_24_tuberack_eppendorf_1.5ml_safelock_snapcap",
+                        "dest_temp_mod_lname":"temperature module gen2",
+                        "do_vortex_pause":"true",
+                        "do_SDS_step":"false",
+                        "do_DNAse_step":"true",
```
Comment (chazmanc) [inline protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:30]: delete

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py [hunk 13]
```diff
@@ -0,0 +1,613 @@
...
+    Add your modules here with:
+
+    module_name = ctx.load_module('{module_loadname}', '{slot number}')
+
+    Note: if you are loading a thermocycler, you do not need to specify
+    a slot number - thermocyclers will always occupy slots 7, 8, 10, and 11.
+
+    For all other modules, you can load them on slots 1, 3, 4, 6, 7, 9, 10.
+
+    '''
```
Comment (chazmanc) [inline protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:147]: delete

Comment (chazmanc) [reply protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:147]: good point by @eskila - can leave in to guide user

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py [hunk 14]
```diff
@@ -0,0 +1,613 @@
...
+    dnase = tuberack.wells_by_name()['A4']
+
+    '''
+
+    lysis_buffer_well = reservoir.wells_by_name()['A1']
+    wash_buffer_well = reservoir.wells_by_name()['A2']
+    elution_buffer_well = reservoir.wells_by_name()['A3']
+    nacl_well = reservoir.wells_by_name()['A4']
+    bead_well = reservoir.wells_by_name()['A5']
+    waste_well = reservoir.wells()[-1]  # Use the last well for waste
```
Comment (chazmanc) [inline protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:455]: This should be `reservoir.wells()[-1].top(-2)`

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py [hunk 15]
```diff
@@ -0,0 +1,613 @@
...
+                                      mix_after=(n_elution_mixes, mix_vol))
+        ctx.comment("\n\nIncubating samples with elution buffer\n")
+        ctx.delay(0, incubation_time)
+        ctx.comment("\n\nAttracting beads to magnets\n")
+        mag_mod.engage()
+        ctx.delay(0, mag_engage_time)
+        ctx.comment("\n\nTransferring protein elution supernatant to" +
+                    "target plate\n")
+        transfer_plate_to_plate(elution_buf_vol, sample_plate, dest_plate)
+        ctx.comment("\n\n~~~~~ End of protocol ~~~~~\n")
```
Comment (chazmanc) [inline protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:613]: remove from if/else

**Review by chazmanc — COMMENTED** [review]
make minor changes and ensure images are ready

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py (hunk 14 above)
Comment (eskila) [reply protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py:423]: I changed the code to conform

File: protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py (hunk 15 above)
Comment (eskila) [reply protocols/generic_protein_purification/generic_protein_purification.ot2.apiv2.py]: Done

**Review by chazmanc — APPROVED** [review]

**Pull request merged**